pyteal/ast/scratch.py

```diff
@@ -69,13 +53,97 @@ def __hash__(self):
         return hash(self.id)
 
 
+Slot.__module__ = "pyteal"
+
+
+class DynamicSlot(Slot):
+    """A Slot whose id is defined dynamically via an expression"""
+
+    def __init__(self, slotIdExpr: Expr) -> None:
+        """Initializes a scratch slot whose id is determined at runtime.
+
+        Args:
+            slotIdExpr: An expression evaluating to TealType.uint64 in the range of 0-255
+            and representing a scratch slot id.
+        """
+        super().__init__()
+
+        assert slotIdExpr is not None, "cannot create a DynamicSlot with no slotIdExpr"
+
+        assert isinstance(
+            slotIdExpr, Expr
+        ), "slotIdExpr must be an Expr but was provided {}".format(type(slotIdExpr))
+
+        self.id: Expr = slotIdExpr
```

Comment:
Code smell: `id` property can either be an `int` or `Expr` based on Slot subclass